app/data_model/questionnaire_store.py

```diff
@@ -2,30 +2,21 @@
 import logging
 
 from app.data_model.answer_store import AnswerStore
-from app.storage.storage_factory import get_storage
 
 logger = logging.getLogger(__name__)
 
 
 class QuestionnaireStore:
 
-    def __init__(self, user_id, user_ik):
-
+    def __init__(self, storage):
         self.metadata = {}
         self.answer_store = AnswerStore()
         self.completed_blocks = []
+        self.storage = storage
 
-        if user_id and user_ik:
-            self.user_id = user_id
-            self.user_ik = user_ik
-        else:
-            raise ValueError("No user_id or user_ik found in session")
-
-        self.storage = get_storage()
-
-        if self.storage.has_data(self.user_id):
-            logger.debug("User %s has previous data loading", user_id)
-            data = self.storage.get(self.user_id, self.user_ik)
+        if self.storage.has_questionnaire_state():
+            self.json_data = self.storage.get_user_data()
```

Comment:
My idea was to make 'initial_data = json.loads(json_data)' and then compare the Python dicts before saving rather than raw strings. And to wrap the loads and dumps calls in 'serialise' and 'deserialise' functions to make it clear what's happening. This would save json.dumps each time the check for changes occurs and also avoid doing comparisons on original JSON string and the newly serialised one (which I wasn't convinced would always serialise to identical strings even if the actual data was the same ~ e.g. The order of JSON objects in the string output)